src/catchup/simulation/ApplyTransactionsWork.cpp

```diff
@@ -543,41 +545,85 @@ ApplyTransactionsWork::onReset()
     // Prepare the HistoryArchiveStream
     mStream = std::make_unique<HistoryArchiveStream>(mDownloadDir, mRange,
                                                      mApp.getHistoryManager());
+    mApplyLedgerWork.reset();
+    mResults.clear();
 }
 
 BasicWork::State
 ApplyTransactionsWork::onRun()
 {
+    auto& lm = mApp.getLedgerManager();
+    auto const& lclHeader = lm.getLastClosedLedgerHeader();
+    auto const& header = lclHeader.header;
+
+    if (mApplyLedgerWork)
+    {
+        mApplyLedgerWork->crankWork();
+        if (mApplyLedgerWork->getState() != State::WORK_SUCCESS)
+        {
+            return mApplyLedgerWork->getState();
+        }
+        else if (mVerifyResults)
+        {
+            checkResults(mApp, header.ledgerSeq, mResults);
+        }
+    }
+
     std::vector<TransactionEnvelope> transactions;
-    std::vector<TransactionResultPair> results;
     std::vector<UpgradeType> upgrades;
-    if (!getNextLedger(transactions, results, upgrades))
+    mResults.clear();
+    if (!getNextLedger(transactions, mResults, upgrades))
     {
         return State::WORK_SUCCESS;
     }
 
-    auto& lm = mApp.getLedgerManager();
-    auto const& lclHeader = lm.getLastClosedLedgerHeader();
-    auto const& header = lclHeader.header;
-
     // When creating SimulationTxSetFrame, we only want to use mMultiplier when
     // generating transactions to handle offer creation (mapping created offer
     // id to a simulated one). When simulating pre-generated transactions, we
     // already have relevant offer ids in transaction results
     auto txSet = std::make_shared<SimulationTxSetFrame>(
-        mNetworkID, lclHeader.hash, transactions, results, mMultiplier);
+        mNetworkID, lclHeader.hash, transactions, mResults, mMultiplier);
 
     StellarValue sv;
     sv.txSetHash = txSet->getContentsHash();
     sv.closeTime = header.scpValue.closeTime + 1;
     sv.upgrades.insert(sv.upgrades.begin(), upgrades.begin(), upgrades.end());
 
     LedgerCloseData closeData(header.ledgerSeq + 1, txSet, sv);
-    lm.closeLedger(closeData);
-    if (mVerifyResults)
-    {
-        checkResults(mApp, header.ledgerSeq, results);
-    }
+    auto applyLedger = std::make_shared<ApplyLedgerWork>(mApp, closeData);
+
+    auto const& ham = mApp.getHistoryArchiveManager();
+    auto const& hm = mApp.getHistoryManager();
+    bool waitForPublish = false;
+    auto condition = [&lm, &ham, &hm, waitForPublish]() mutable {
```

Comment:
can you explain why we need this? I think it was because we would create a huge backlog of things to publish?

Reply:
Yes, as discussed earlier, publish can't keep up during replay.